rangeserver/src/cache/memtabledb.rs

```diff
@@ -0,0 +1,643 @@
+                oldest_epoch: 0,
+                // TODO: init with capacity
+                entries: OrderedSkipList::<MemTableEntry>::new(),};
+            mt_db.memtables.push(mt);
+        }
+        // change the status of first memtable as mutable
+        mt_db.memtables[mt_db.mutable_id].status = MemTableStatus::MUTABLE;
+        mt_db
+    }
+
+    async fn upsert_or_delete(
+        &mut self,
+        key: bytes::Bytes,
+        epoch: u64,
+        deleted: bool,
+        val: Option<bytes::Bytes>) -> Result<(), Error> {
+
+            // sanity checks
+            assert!(self.mutable_id >= 0 && self.mutable_id < self.rc_options.num_write_buffers, 
+                "mutable_id is out of range ({})", self.mutable_id);
+            
+            let mut val_match = Bytes::new();
+            let mut val_len = 0; 
+            match val{
+                Some(v) => {
+                    val_len = v.len();
+                    val_match = v;
+                },
+                None => assert!(deleted == true, "upsert api called without value argument"),
+            }
+
+            let entry_size: u64 = key.len() as u64 + (val_len as u64) + (mem::size_of::<u64>() as u64) + (mem::size_of::<bool>() as u64);
+            let entry = MemTableEntry { key: key, val: val_match, epoch: epoch, deleted: deleted };
+            
+            assert!(self.memtables[self.mutable_id].status == MemTableStatus::MUTABLE, 
+                "active memtable is not mutable ({}) ({:?})", self.mutable_id, self.memtables[self.mutable_id].status);
+
+            // println!("Memtable size {} entry size {}", memtable_wg.size_in_bytes, entry_size);
+            if self.memtables[self.mutable_id].size_in_bytes + entry_size > self.rc_options.write_buffer_size {
+                let old_id= self.mutable_id;
+                let new_id = (self.mutable_id+1) % self.rc_options.num_write_buffers;
+
+                assert!(self.memtables[new_id].status == MemTableStatus::FLUSHED, 
+                    "no free memtables {} {:?}", new_id, self.memtables[new_id].status);
+                // Err(Error::OutOfSpace)
+
+                self.mutable_id = new_id;
+                self.memtables[new_id].status = MemTableStatus::MUTABLE;
+                self.memtables[old_id].status = MemTableStatus::IMMUTABLE;
+                self.memtables[new_id].oldest_epoch = epoch;
+                println!("Switched memtable from {} to {}", old_id, new_id);
+
+                // trigger garbage collection on flushed memtable
+                match self.gc_callback {
+                    Some(cb) => cb(self.memtables[self.mutable_id].latest_epoch),
+                    None => println!("INFO: GC callback is not set {:?}", self.gc_callback),
+                }
+            }
+            self.memtables[self.mutable_id].entries.insert(entry);
+            self.memtables[self.mutable_id].size_in_bytes += entry_size;
+            self.memtables[self.mutable_id].latest_epoch = epoch;
```

Comment:
Any chance incoming `epoch` would be earlier than a previously seen one in a race situation?

Reply:
Range server uses the cache in single writer multiple reader mode. So the race you mentioned shouldn't happen. But I will go ahead and assign the max of latest epoch and epoch.